### DISCUSSION TITLE
Bug fixes and new features

### DISCUSSION THREAD

**Discussion by Tirlititi** [Albeoris/Memoria#169] [General]
* Added AnimationClip importers (both serialized binary files and Json-formatted animations).
* Added a way to turn a character into any enemy in a battle, with BattleUnit.ChangeToMonster.
* Bugfixes.

<hr /><em>This discussion was created from the release <a href='https://github.com/Albeoris/Memoria/releases/tag/v2021.05.10'>Bug fixes and new features</a>.</em>

**Comment by Tirlititi** [comment]
I upload here several files that can be useful for features related to this release: an example of a JSON animation file, a short description of "DictionaryPatch.txt" features, a database of the game's model IDs and animations and a database of the battle IDs.

[JSONAnimationSample.zip](https://github.com/Albeoris/Memoria/files/6454133/JSONAnimationSample.zip)
[Dictionary Patch manual.txt](https://github.com/Albeoris/Memoria/files/6454136/Dictionary.Patch.manual.txt)
[FF9 Model Database.xls](https://github.com/Albeoris/Memoria/files/6454141/FF9.Model.Database.xls)
[Battle Database.xls](https://github.com/Albeoris/Memoria/files/6454144/Battle.Database.xls)

Also, a modified version of the Ultima SFX that can be used for player abilities can be found [there](https://www.dropbox.com/s/01qjh0azopnpvsm/PC_SetupUltima.zip?dl=1).
And a mod letting the player play with Kuja as a main character (with many glitches because animations were not worked on), using mainly DictionaryPatch features, can be found [there](https://www.hiveworkshop.com/attachments/playaskuja-zip.377400/).

**Comment by FlameHorizon** [comment]
Is there a way to modify treasures which are coming from Chocobo Hot & Cold minigame?

**Reply by Tirlititi** [reply]
...
I've misread your question and started answering "How to change treasures given by Chocographs?". Nevermind, I'll give both answers.

-------------------------------------------
**How to change treasures given by Chocographs?**

That's part of the game script system. In Hades Workshop, go to the "Environment -> World Map" panel, edit the script and search for a function looking like this one:
![HadesWorkshop_ChocographRewards](https://user-images.githubusercontent.com/20627826/148593868-558b1777-a91f-40cc-b511-79ce90c79954.png)

Each chocograph sets 8 variables in the following order:
[1st item ID, 1st item amount, 2nd item ID, 2nd item amount, 3rd item ID, 3rd item amount, 4th item ID, 4th item amount]
In the picture above, I have the caret on the 1st item ID of the 1st chocograph, which is 239 and you can see on the left that it corresponds to an Elixir. The 1st chocograph gives 2 Elixirs, 3 Hi-Potions, 4 Ethers and 2 Germinas Boots.

Note that the 4th item ID is a variable of type "Int16" instead of "UInt8". That is because the 4th item can be a Tetra Master card and thus get a value higher than 255. You'll have troubles if you try to give a card in any other slot than the 4th.

In order to modify the chocograph's reward, change the figure there, then parse and confirm. You need to do that for all the different World Maps (except the "World Map/Event" ones).

The rewards of bubbles and mountain cracks can be found in "World Chocobo" functions (for the 1st World Map, it is the functions "World_Chocobo_15" and "World_Chocobo_16") and they work similarly.

-------------------------------------------
**How to change treasures in Hot & Cold minigame?**

That also is a game script and must be modified by Hades Workshop. That one is in the "Environment -> Fields" panel instead (it takes longer to load). Search for the Hot & Cold fields and use "Edit Script". The function you're looking for is "Function Barrel_41" (at least for the Chocobo Forest). It starts with the following:

    Function Barrel_41
        set VAR_GlobUInt16_47 ^= ( 1 << VAR_LocUInt8_1 )
        set VAR_LocUInt8_12 = 0
        set VAR_LocUInt8_14 = 0
        if ( Chocobo_ChocographKnown == 0 ) {
            set VAR_LocUInt8_16 = 48
        }
        ...


The first big "if" block is all about getting a chocograph, so you want to scroll down a bit and reach this part:

    if ( VAR_LocUInt8_12 == 0 ) {
        if ( ( ( !GetRandom ) && ( GetRandom < ( 64 + ( 10 * Chocobo_GoodTreasureBonus ) ) ) ) == 1 ) {
            set VAR_GlobUInt8_53 = 5
            set VAR_LocUInt8_13 = 16
            switch 8 ( GetRandom % 9 ) from 0 {
            case +0:
                set VAR_LocUInt16_20 = 218 // Cachusha
                break
            case +1:
                set VAR_LocUInt16_20 = 219 // Barette
                break
            case +2:
                set VAR_LocUInt16_20 = 220 // Extension
                break
            case +3:
                set VAR_LocUInt16_20 = 86 // Rising Sun
                break
            case +4:
                set VAR_LocUInt16_20 = 199 // Anklet
                break
            case +5:
                set VAR_LocUInt16_20 = 87 // Wing Edge
                break
            case +6:
                set VAR_LocUInt16_20 = 11000 // 10,000 gil
                break
            case +7:
                set VAR_LocUInt16_20 = 1001 // 1 gil
                break
            default:
                set VAR_LocUInt16_20 = 85 // Pinwheel
                break
            }
        } else {
            ...

Each number value in these switches are the item ID of the treasure you dig. I've added the item name in the code above but you can see what item corresponds to what ID by selecting the number and looking at the left part of the screen (you need to have first gone in the "Items" panel of Hades Workshop to get this help).

You can see these are the rarest items you can dig. The more common items can be found a bit below in similar lists. Change the item ID there, then parse and confirm.

I've made a slightly more readable version of the Hot & Cold minigame reward functions. [You can read it there](https://www.dropbox.com/s/9j71di7ijcfpzii/Script_Hot%26ColdReward.txt?dl=0) and use it to find your way in the script.
